cumulus/polkadot-parachain/src/service.rs

```diff
@@ -1530,6 +1723,157 @@ where
+		para_id,
+		|_| Ok(RpcModule::new(())),
+		aura_build_import_queue::<_, AuraId>,
+		|client,
+		 block_import,
+		 prometheus_registry,
+		 telemetry,
+		 task_manager,
+		 relay_chain_interface,
+		 transaction_pool,
+		 sync_oracle,
+		 keystore,
+		 relay_chain_slot_duration,
+		 para_id,
+		 collator_key,
+		 overseer_handle,
+		 announce_block,
+		 backend| {
+			let relay_chain_interface2 = relay_chain_interface.clone();
+
+			let collator_service = CollatorService::new(
+				client.clone(),
+				Arc::new(task_manager.spawn_handle()),
+				announce_block,
+				client.clone(),
+			);
+
+			let spawner = task_manager.spawn_handle();
+
+			let proposer_factory = sc_basic_authorship::ProposerFactory::with_proof_recording(
+				spawner,
+				client.clone(),
+				transaction_pool,
+				prometheus_registry,
+				telemetry.clone(),
+			);
+
+			let collation_future = Box::pin(async move {
+				// Start collating with the `shell` runtime while waiting for an upgrade to an Aura
+				// compatible runtime.
+				let mut request_stream = cumulus_client_collator::relay_chain_driven::init(
+					collator_key.clone(),
+					para_id,
+					overseer_handle.clone(),
+				)
+				.await;
+				while let Some(request) = request_stream.next().await {
+					let pvd = request.persisted_validation_data().clone();
+					let last_head_hash =
+						match <Block as BlockT>::Header::decode(&mut &pvd.parent_head.0[..]) {
+							Ok(header) => header.hash(),
+							Err(e) => {
+								log::error!("Could not decode the head data: {e}");
+								request.complete(None);
+								continue
+							},
+						};
+
+					// Check if we have upgraded to an Aura compatible runtime and transition if
+					// necessary.
+					if client
+						.runtime_api()
+						.has_api::<dyn AuraApi<Block, AuraId>>(last_head_hash)
+						.unwrap_or(false)
+					{
+						// Respond to this request before transitioning to Aura.
+						request.complete(None);
+						break
+					}
+				}
+
+				// Move to Aura consensus.
+				let slot_duration = match cumulus_client_consensus_aura::slot_duration(&*client) {
+					Ok(d) => d,
+					Err(e) => {
+						log::error!("Could not get Aura slot duration: {e}");
+						return
+					},
+				};
+
+				let proposer = Proposer::new(proposer_factory);
+
+				let params = AuraParams {
+					create_inherent_data_providers: move |_, ()| async move { Ok(()) },
+					block_import,
+					para_client: client.clone(),
+					para_backend: backend,
+					relay_client: relay_chain_interface2,
+					code_hash_provider: move |block_hash| {
+						client.code_at(block_hash).ok().map(|c| ValidationCode::from(c).hash())
+					},
+					sync_oracle,
+					keystore,
+					collator_key,
+					para_id,
+					overseer_handle,
+					slot_duration,
+					relay_chain_slot_duration,
+					proposer,
+					collator_service,
+					authoring_duration: Duration::from_millis(1500),
+					// TODO: is the below needed?
+					//collation_request_receiver: Some(request_stream),
```

Comment:
Is this still needed here (see `start_asset_hub_node`), or is it only for transitioning to aura above?

Reply:
I see that the lookahead implementation doesn't use the [relay chain driven collation request stream](https://github.com/paritytech/polkadot-sdk/blob/f7306d327e0eee0a641ac43628ab525b0096ba7c/cumulus/client/consensus/aura/src/collators/basic.rs#L122), but instead uses [`cumulus_client_collator::initialize_collator_subsystems`](https://github.com/paritytech/polkadot-sdk/blob/f7306d327e0eee0a641ac43628ab525b0096ba7c/cumulus/client/consensus/aura/src/collators/lookahead.rs#L148), which still [sends a message](https://github.com/paritytech/polkadot-sdk/blob/f7306d327e0eee0a641ac43628ab525b0096ba7c/cumulus/client/collator/src/lib.rs#L248) which will [error as a double initialization](https://github.com/paritytech/polkadot-sdk/blob/f7306d327e0eee0a641ac43628ab525b0096ba7c/polkadot/node/collation-generation/src/lib.rs#L136). To fix this, you will probably have to create a special message to override an existing config with the lookahead specific one.

Reply:
⭐ 

Thanks @georgepisaltu for walking me through these issues - super helpful! What do you think would be a reasonable change here? I'm thinking we could add a new parameter to `initialize_collator_subsystems` like `reinitialize: bool` and if it's true, send a new message like `CollationGenerationMessage::Reinitialize` instead of `CollationGenerationMessage::Initialize` which doesn't have an initialization limit error. Do you think this makes sense or is there a better way to do it?

Reply:
> we could add a new parameter to `initialize_collator_subsystems` like `reinitialize: bool`

Sounds good to me.

> send a new message like `CollationGenerationMessage::Reinitialize` instead of `CollationGenerationMessage::Initialize` which doesn't have an initialization limit error

Also sounds good. The other option I see is to have a new `override` or `force` parameter in the `Initialize` message, which, if set, ignores any existing config and replaces it with the new one.

But I think creating a new message for this is better because it's backwards compatible.